### PR TITLE
Prepublication

### PR DESCRIPTION
**Pull request by vladdez**
- #244 Added benchmarking with comparison of UM with MNE and MATLAB to measure speed of creating and updating figures. 
- adding supportive axes for `plot_topoplot`, `plot_topoplotseries`, `plot_butterfly` so now it is possible for users to change interpolations and axes of subplots
- adding Observables into `plot_topoplot`
- #245 was solved by adding `visual.colormap`
- #248 
- #240
- #257 
- #261 - issue with `with_theme`
- `Contribute` page added
- issues templates added
- upgrade Color and ColorTypes

### CODE REVIEW

File: src/plot_butterfly.jl [hunk 1]
```diff
@@ -154,8 +160,8 @@ function plot_butterfly!(
 end
 
 
-# topopositions_to_color = colors?
-function topoplot_legend(axis, topomarkersize, unique_val, colors, all_positions)
+# should we make this adjustable?
```
Comment (behinger) [inline src/plot_butterfly.jl:163]: what exactly is the question here?

File: src/plot_erp.jl [hunk 2]
```diff
@@ -31,6 +31,10 @@ Plot an ERP plot.
     Enable or disable legend.\\
 - `layout.show_legend = true`\\
     Enable or disable legend and colorbar.\\
+- `mapping = (;)`\\
+    Specify `color`, `col` (column), `linestyle`, `group`.\\
+- `visual.color = Makie.wong_colors, visual.colormap = :roma`\\
+    For categorical color use `visual.color`, for contionous - `visual.colormap`.\\
```
Comment (behinger) [inline src/plot_erp.jl:37]: typo in continuous

File: src/plotconfig.jl [hunk 3]
```diff
@@ -213,6 +213,7 @@ function PlotConfig(T::Val{:erp})
             xtickformat = "{:.1f}",
         ),
         colorbar = (; label = "", flipaxis = true),
+        visual = (; colormap = :roma, color = Makie.wong_colors()),
```
Comment (behinger) [inline src/plotconfig.jl:216]: do we really need this? `color` specifies only the Makie defaults anyway, so then you could never overwrite them with `with_theme` which would be a shame

File: src/plot_butterfly.jl (hunk 1 above)
Comment (vladdez) [reply src/plot_butterfly.jl:163]: already implemented

File: src/plot_erp.jl [hunk 4]
```diff
@@ -142,7 +146,18 @@ function plot_erp!(
     plot_equation = basic * mapp
 
     f_grid = f[1, 1] = GridLayout()
-    drawing = draw!(f_grid, plot_equation; axis = config.axis)
+
+    color_mapper =
+        isa(config.mapping.color, Symbol) ? config.mapping.color : config.mapping.color[1]
```
[the remark this answers is not in the document]
Comment (vladdez) [reply src/plot_erp.jl:152]: you need to check first that `config.mapping.color[1] `exist

File: src/plotconfig.jl (hunk 3 above)
Comment (vladdez) [reply src/plotconfig.jl:216]: okay, that actually happening. Where i can catch color form with_theme in the function?

File: docs/literate/intro/speed.jl [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+simulated_epochs = PyMNE.EvokedArray(Py(dat[:, :, 1]), info)
+
+@benchmark simulated_epochs.plot_topomap(1:50)
+
+# MATLAB
+#
+# At present, it is not possible to run MATLAB within a Julia environment on GitHub. As a result, we can only provide a screenshot of the program execution.
+
+# ```@raw html
+# <img src="../../../assets/MATLAB_benchmarking.png" align="middle"/>
```
Comment (behinger) [inline docs/literate/intro/speed.jl:88]: here you can use the fix we introduced today

File: docs/literate/intro/speed.jl [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+end
+
+# MNE
+easycap_montage = PyMNE.channels.make_standard_montage("standard_1020")
+ch_names = pyconvert(Vector{String}, easycap_montage.ch_names)[1:64]
+info = PyMNE.create_info(PyList(ch_names), ch_types = "eeg", sfreq = 1)
+info.set_montage(easycap_montage)
+simulated_epochs = PyMNE.EvokedArray(Py(dat[:, :, 1]), info)
+
+@benchmark simulated_epochs.plot_topomap(1:50)
```
Comment (behinger) [inline docs/literate/intro/speed.jl:81]: there is no averaging involved here, probably doesn change much, but a disclamer should be added

File: docs/literate/intro/speed.jl [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+dat, positions = TopoPlots.example_data()
+df = UnfoldMakie.eeg_array_to_dataframe(dat[:, :, 1], string.(1:length(positions)));
+
+# # Topoplots
+
+# UnfoldMakie.jl
+@benchmark plot_topoplot(dat[:, 320, 1]; positions = positions)
+
+# UnfoldMakie.jl with DelaunayMesh
+@benchmark plot_topoplot(
```
Comment (behinger) [inline docs/literate/intro/speed.jl:28]: I've seen that there is little difference, maybe remove delaunay mesh here?

File: docs/literate/intro/speed.jl [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+ch_names = pyconvert(Vector{String}, easycap_montage.ch_names)[1:64]
+info = PyMNE.create_info(PyList(ch_names), ch_types = "eeg", sfreq = 1)
+info.set_montage(easycap_montage)
+simulated_epochs = PyMNE.EvokedArray(Py(dat[:, :, 1]), info)
+
+@benchmark simulated_epochs.plot_topomap(1:50)
+
+# MATLAB
+#
+# At present, it is not possible to run MATLAB within a Julia environment on GitHub. As a result, we can only provide a screenshot of the program execution.
```
Comment (behinger) [inline docs/literate/intro/speed.jl:85]:
```suggestion
# It is not easy to run MatLab as a GitHub-Action. We therefore provide execution times from three consecutive executions on comparatively fast machine (64 cores, from 2020)
```

File: docs/literate/intro/speed.jl [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    plot_topoplot!(f[1, 1], dat_obs, positions = positions)
+    record(f, "topoplot_animation_UM.gif", timestamps; framerate = framerate) do t
+        dat_obs[] = @view(dat[:, t, 1])
+    end
+end
+
+# ![](topoplot_animation_UM.gif)
+
+# MNE with .gif
+@benchmark begin
```
Comment (behinger) [inline docs/literate/intro/speed.jl:113]: MNE still is only contour

File: src/eeg_series.jl [hunk 10]
```diff
@@ -147,59 +160,59 @@ function eeg_topoplot_series!(
...
         @assert isa(interactive_scatter, Observable)
     end
     for r = 1:length(select_row)
         for c = 1:length(select_col)
-            ax = Axis(fig[:, :][r, c]; axis_options...)
             df_single =
                 topoplot_subselection(data_mean, col, row, select_col, select_row, r, c)
+            single_y = @lift($df_single[:, y])
+            if isempty(to_value(single_y))
+                break
```
Comment (behinger) [inline src/eeg_series.jl:176]: why is this needed? please add a comment explaining it

File: src/eeg_series.jl [hunk 11]
```diff
@@ -147,59 +160,59 @@ function eeg_topoplot_series!(
...
+            end
+
+            ax = Axis(     #here we loose 30 seconds
```
Comment (behinger) [inline src/eeg_series.jl:179]: what does loose 30seconds mean?

File: src/plot_erp.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (behinger) [reply src/plot_erp.jl:152]: even if you check this, wouldnt my solution be a bit cleaner?

File: src/plotconfig.jl (hunk 3 above)
Comment (behinger) [reply src/plotconfig.jl:216]: why do you need to catch it?

File: src/plotconfig.jl [hunk 12]
```diff
@@ -253,7 +254,7 @@ function PlotConfig(T::Val{:erpimage})
         cfg;
         layout = (; use_colorbar = true),
         colorbar = (; label = "Voltage [µV]"),
-        axis = (xlabel = "Time [s]", ylabel = "Trials"),
+        axis = (xlabel = "Time samples", ylabel = "Trials"),
```
Comment (behinger) [inline src/plotconfig.jl:257]: this is not correct - if you call plot_erpimage with a dataframe, you should get time[s] if you call it with a matrix, you should get samples. Therefore, yo have to update this in the conversion function

File: src/supportive_defaults.jl [hunk 13]
```diff
@@ -49,6 +55,45 @@ function supportive_defaults(cfg_symb::Symbol)
...
+        # plot_topoplot 
+    elseif cfg_symb == :topo_default_single
+        return (;
+            width = Relative(1),
+            height = Relative(1),
+            halign = 0.05,
+            valign = 0.95,
+            aspect = 1,
+        )
+    elseif cfg_symb == :topo_attributes_default
```
Comment (behinger) [inline src/supportive_defaults.jl:67]: - why not topo_default_attributes?
- is the default even necessary throughout here?
- can't we add these to `axis` in plot_config - at least for e.g. topoplot series?

**Review by behinger — COMMENTED** [review]
please please please make smaller PR's in the future - one per feature / change :)

File: src/plotconfig.jl (hunk 3 above)
Comment (vladdez) [reply src/plotconfig.jl:216]: to be able to prioritize it in the code, for instance

File: docs/literate/intro/speed.jl (hunk 9 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:113]: yes, but I dont have any clue why is it like that

File: src/eeg_series.jl (hunk 10 above)
Comment (vladdez) [reply src/eeg_series.jl:176]: it is basically end of loop
i will put a comment

File: docs/literate/intro/speed.jl (hunk 6 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:81]: averaging of what?

File: docs/literate/intro/speed.jl (hunk 7 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:28]: 300 ms
Good question, we can remove it or also inlude mne with linear interpolation

File: docs/literate/intro/speed.jl (hunk 8 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:85]: what is machine here? Matlab version? isnt it 2021a (according to wiki)?

File: src/eeg_series.jl (hunk 11 above)
Comment (vladdez) [reply src/eeg_series.jl:179]: probabaly i meant that creating axis requires 30 ms

File: src/plotconfig.jl (hunk 12 above)
Comment (vladdez) [reply src/plotconfig.jl:257]: yeah, but  we currently do not accept DataFrames in this function and as is understood it is not really necessary

File: docs/literate/intro/speed.jl (hunk 5 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:88]: no, it didnt work for some strange reason
I will keep it until I will get why

File: src/supportive_defaults.jl (hunk 13 above)
Comment (vladdez) [reply src/supportive_defaults.jl]: good quesion, I will play with configs later

File: docs/literate/intro/speed.jl (hunk 6 above)
Comment (behinger) [reply docs/literate/intro/speed.jl:72]: averaging over time-samples. in unfoldmakie we ask for 50 topoplots, where each is a mean over 4 samples, whereas in plot_topopmap you ask for the first 50 samples to be visualized.

this is all as far as I remember, and we should just make a note of this. I dont expect taking the mean will take that much time to explain the speed advantage of MNE

File: docs/literate/intro/speed.jl (hunk 9 above)
Comment (behinger) [reply docs/literate/intro/speed.jl:103]: if you run this code on your computer, is it also only contour?

File: src/plotconfig.jl (hunk 3 above)
Comment (behinger) [reply src/plotconfig.jl:216]: I dont understand your answer. What is lost if we simply remove the color here? Then users could set it with `with_theme`

File: src/plotconfig.jl (hunk 12 above)
Comment (behinger) [reply src/plotconfig.jl:257]: ah sorry - but you can call it with `times, data` - maybe then switch to `Time` without unit, because we dont know the unit?

File: docs/literate/intro/speed.jl (hunk 8 above)
Comment (behinger) [reply docs/literate/intro/speed.jl]: (if you ask a question, but also resolve the thread, I dont knw if you want an answer or not)

Machine = computer, 2020 is the time I bought the computer; you could bette rlook up the processor we have, there is probably a command line command for it :)

File: docs/literate/intro/speed.jl (hunk 9 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:103]: yes!

File: docs/literate/intro/speed.jl (hunk 6 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl:72]: addressed

File: docs/literate/intro/speed.jl (hunk 9 above)
Comment (behinger) [reply docs/literate/intro/speed.jl:103]: there is a chance that Pillow requires ffmpeg after all.

have you tried `using CondaPkg; CondaPkg.add("ffmpeg")` alredy to install ffmpeg?

File: src/supportive_defaults.jl (hunk 13 above)
Comment (vladdez) [reply src/supportive_defaults.jl]: "It is not easy to run Matlab as a GitHub action. Therefore, we provided execution times of three consecutive executions."
is such text okay or something else needed?

File: src/plotconfig.jl (hunk 12 above)
Comment (vladdez) [reply src/plotconfig.jl:257]: okay i will delete all units

File: docs/literate/intro/speed.jl (hunk 8 above)
Comment (vladdez) [reply docs/literate/intro/speed.jl]: "It is not easy to run Matlab as a GitHub action. Therefore, we provided execution times of three consecutive executions."
is such text okay or something else neede

Comment (behinger) [reply docs/literate/intro/speed.jl]: I would mention that the times are on a different computer, and you cant compare them directly - likely the matlab numbers are too fast compared to what is printed via github

File: src/plotconfig.jl (hunk 12 above)
Comment (behinger) [reply src/plotconfig.jl]: why? in some cases we know the units, no? in other plots you also assume units, so to some degree I think this is fine?

File: src/supportive_defaults.jl (hunk 13 above)
Comment (vladdez) [reply src/supportive_defaults.jl]: default is necessary (I guess) because we want to control interpolation
I think the main reason not to include it to config.axis is that by ?Axis you will not get documentation for it. You need ?Topoplot.topoplot

File: src/plotconfig.jl (hunk 12 above)
Comment (vladdez) [reply src/plotconfig.jl]: I think time units are not necessary (could be misleading), nut about mV i am not sure

File: src/supportive_defaults.jl (hunk 13 above)
Comment (vladdez) [reply src/supportive_defaults.jl]: also you cannot just do `Axis(f[1, 1]; config.axis...)` with interpolation, because it doesn't recognize such attributes

Comment (vladdez) [reply src/supportive_defaults.jl]: so I would vote for keeping it
